Issue #396: Certain ImageDescriptor classes should be adaptable for...

Dani-Hub opened the pull request:
...some internal properties

- Let FileImageDescriptor and URLImageDescriptor implement IAdaptable and provide adaption to ImageFileNameProvider (for testing purposes)
- Add unit tests that attempt to retrieve the underlying ImageFileNameProvider and validate the expected xPath results.

Signed-off-by: Daniel Krügler <daniel.kruegler@gmail.com>

Dani-Hub committed:
Issue #396: Certain ImageDescriptor classes should be adaptable for...

...some internal properties

- Let FileImageDescriptor and URLImageDescriptor implement IAdaptable
and provide adaption to ImageFileNameProvider (for testing purposes)
- Add unit tests that attempt to retrieve the underlying
ImageFileNameProvider and validate the expected xPath results.

Signed-off-by: Daniel Krügler <daniel.kruegler@gmail.com>

Dani-Hub commented:
@laeubi Can you please help me understanding the cause of the test failures? I have problems determining the actual error, albeit it seems to be related to my PULL request, but to which extend?

laeubi commented:
It seems an assertion failed:

https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-397/1/testReport/junit/org.eclipse.jface.tests.images/FileImageDescriptorTest/testImageFileNameProviderGetxPath/

I would recommend to enhance the test, e.g. with an meaningful message at the assertion.

laeubi commented on tests/org.eclipse.jface.tests/src/org/eclipse/jface/tests/images/FileImageDescriptorTest.java:
```diff
@@ -174,4 +176,22 @@ public void testGetxPath150() {
+				"/icons/imagetests/rectangular-57x16.png");
+		assertTrue("FileImageDescriptor does not implement IAdaptable", descriptor instanceof IAdaptable);
+		IAdaptable adaptable = (IAdaptable) descriptor;
+		ImageFileNameProvider fileNameProvider = adaptable.getAdapter(ImageFileNameProvider.class);
```
Usually the pattern is to use:

Adapters.getAdapter

instead of checking for adaptable directly.

Dani-Hub replied:
I agree that this is what should be done in production code, but this unit test particularly wants to tests the IAdaptable inheritance.

laeubi replied:
Does it really matter "how" it is adapted? I think the important part is *that* it could be adapted....

Dani-Hub replied:
Well, it does not matter to me, certainly. My test was constructed based on the use case in [MenuHelper.java#L491](https://github.com/eclipse-platform/eclipse.platform.ui/blob/849e023520392865995477474ef731f5302edbda/bundles/org.eclipse.ui.workbench/Eclipse%20UI/org/eclipse/ui/internal/menus/MenuHelper.java#L491) where also such a concrete tests exists. I have no strong feelings about this aspect. If replacing the concrete test for IAdaptable inheritance and its usage by a call to `org.eclipse.core.runtime.Adapters.adapt(Object sourceObject, Class<T> adapter)` or `org.eclipse.core.runtime.Adapters.of(Object sourceObject, Class<T> adapter)` would increase consensus, let me know. @laeubi please respond.

Dani-Hub commented:
> It seems an assertion failed:
> 
> https://ci.eclipse.org/platform/job/eclipse.platform.ui/job/PR-397/1/testReport/junit/org.eclipse.jface.tests.images/FileImageDescriptorTest/testImageFileNameProviderGetxPath/

Thanks!

> I would recommend to enhance the test, e.g. with an meaningful message at the assertion.

Well, I just copied the existing style of the other tests ;-)

OK, I found the problem and it was actually caused by an error of my previous #376 implementation - and this new tests demonstrates the importance of such a test - thanks for reminding to make this testable!

As a matter of fact, it also turns out, that the _original_ implementation also didn't work correctly, so I will now fix both the original as well as my own extension.

laeubi commented:
> Well, I just copied the existing style of the other tests ;-)

Don't do that! :-D

Actually it is always good to have meaningful assertion with a message that tells what has failed for "simple" asserts like numbers or boolean as its often hard to guess what "expected 2 but got 3" means or "assertion failed" like here.

Dani-Hub committed:
Fix ImageFileNameProvider implementations

- ImageFileNameProvider proceeds searching when xName or xPath approach
fails
- Optimization for 100% zoom to not perform unnecessary work with xName
or xPath
- Extend unit tests by adding negative tests for unexpected zoom values
and by adding xName tests
- Improve unit tests by providing specific assertion messages

Signed-off-by: Daniel Krügler <daniel.kruegler@gmail.com>

Dani-Hub commented:
@laeubi All tests pass now, how do you like it?

laeubi commented on bundles/org.eclipse.jface/src/org/eclipse/jface/resource/FileImageDescriptor.java:
```diff
@@ -295,4 +306,13 @@ String getFilePath(String name, boolean logIOException) {
+	public <T> T getAdapter(Class<T> adapter) {
+		if (adapter == ImageFileNameProvider.class) {
+			// Support testing ImageFileNameProvider characteristics, see #396
+			return adapter.cast(new ImageProvider());
```
Could an image provide maybe only instantiated once?

Dani-Hub replied:
Well, it could, but that's not how the FileImageDescriptor class is currently implemented. You should notice that it always generates a fresh new ImageProvider when it's required. And note also that the sole purpose is (currently) for testing. I would suggest that we should create a separate issue if we want to have only a single ImageFileNameProvider instance within the ImageDescriptor, because that would be a specific optimization approach, unrelated to the new adapter code. You can count on me to volunteer for the issue and taking care of a PULL request.

Dani-Hub replied:
A corresponding issue has been created, see #398.

laeubi commented on bundles/org.eclipse.jface/src/org/eclipse/jface/resource/URLImageDescriptor.java:
```diff
@@ -344,4 +351,13 @@ private static URL getURL(String urlString) {
+	public <T> T getAdapter(Class<T> adapter) {
+		if (adapter == ImageFileNameProvider.class) {
+			// Support testing ImageFileNameProvider characteristics, see #396
+			return adapter.cast(new URLImageFileNameProvider(url));
```
Same here, why always a new instance?

Dani-Hub replied:
See above
